### PR TITLE
(For Review) Manual Mod Updating

### PR DESCRIPTION
**Pull request by Bletch1971**
Part 1 - I have added the functionality to manually update the mod files
based on the ids specified in the ASM settings.

The mod update has been added to the ServerUpdate button and also a new
Mod Update button (near the mod id field).

### CODE REVIEW

**Comment by Bletch1971** [comment]
Cliff, A problem exists with the enabling of the new Mod Update button.
It is always disabled when opening ASM. The button code has been copied from the Server Update button but I am unable to get it to enable the same way.

**Comment by ChronosWS** [comment]
DataContexts are inherited.  It looks like the container for the button (the ScrollViewer on line 694) is setting the DataContext to Profile, so Runtime isn't available there.  You will need to explicitly reference Runtime using another binding path (like DataContext.Runtime, ElementName=SettingsControl).  If you look in the debug window when running the application, you should see a BindingError, which will go away when you have solved this.

**Comment by Bletch1971** [comment]
Thank you so much

On 13 Jan 2016, at 5:37 PM, Cliff Hudson notifications@github.com wrote:

DataContexts are inherited. It looks like the container for the button (the ScrollViewer on line 694) is setting the DataContext to Profile, so Runtime isn't available there. You will need to explicitly reference Runtime using another binding path (like DataContext.Runtime, ElementName=SettingsControl). If you look in the debug window when running the application, you should see a BindingError, which will go away when you have solved this.

**Comment by Bletch1971** [comment]
OK, I think I have this nailed now, thanks to you.

Manual mod updating has been implemented into ASM. I will start to look at getting the automatic process done too.

**Comment by Bletch1971** [comment]
I you feel this needs to be tested a bit, then I know that a few of the ASM forum members might help out. We have a few regulars, like Beals, Anoobis, Kane.

**Comment by Bealze** [comment]
is there  a way i can pull a zip with these changes ill be glade to run it

**Comment by Bletch1971** [comment]
Are you able to compile the code, or do you mean a zip of the executables?

**Comment by Bletch1971** [comment]
I am working to implement in the mod update to include SotF (when selected), TotalConversions and also Map mods.

**Comment by Bletch1971** [comment]
Checking this in, so I can work on the 232 updates.

Still to do, when starting the server, perform a check that ALL mods have been downloaded, including SotF.

File: ARK Server Manager/Lib/ModUpdater/ModCopy.cs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+                        if (metaInformation.Count == 0 && flag)
+                            metaInformation["ModType"] = "1";
+
+                        WriteModFile($"{serverModPath}.mod", id, metaInformation, list);
+                    }
+                }
+                return true;
+            }
+            catch
```
Comment (ChronosWS) [inline ARK Server Manager/Lib/ModUpdater/ModCopy.cs]: Is it possible to get exceptions in the normal course, and should we log something or bubble up an error message?  Unconditional catches are usually an indication of missing failure case handling.

File: ARK Server Manager/Lib/ModUpdater/ModCopy.cs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Ionic.Zlib;
+
+namespace ARK_Server_Manager.Lib
+{
+    public static class ModCopy
+    {
+        private class FCompressedChunkInfo
```
Comment (ChronosWS) [inline ARK Server Manager/Lib/ModUpdater/ModCopy.cs]: Does this code all come from another source?  Are we attributing it?

File: ARK Server Manager/Lib/ModUpdater/ModUpdater.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                    copyProcess.ErrorDataReceived += (s, e) => copyTS.TrySetException(new Exception(e.Data));
+                    if (!copyTS.Task.Result)
+                        return Task.FromResult<bool>(false);
+                }
+
+                return Task.FromResult<bool>(true);
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine($"ModUpdater.UpgradeAsync - {ex.Message}");
```
Comment (ChronosWS) [inline ARK Server Manager/Lib/ModUpdater/ModUpdater.cs:254]: Can we log this to the debug log as well?

File: ARK Server Manager/Lib/ModUpdater/ModUpdater.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                ModCopy.GetModDetails(serverInstallDirectory, id, out metaInformation, out mapNames);
+
+                if (mapNames != null && mapNames.Count > 0)
+                    return Task.FromResult<string>(mapNames[0]);
+
+                return Task.FromResult<string>(null);
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine($"ModUpdater.GetMapName - {ex.Message}");
```
Comment (ChronosWS) [inline ARK Server Manager/Lib/ModUpdater/ModUpdater.cs]: And here as well.

File: ARK Server Manager/Lib/Model/ServerRuntime.cs [hunk 5]
```diff
@@ -380,23 +393,117 @@ public static void SendStop(Process proc)
...
-        }       
+        }
+
+        public async Task<UpdateResult> UpgradeModsAsync(CancellationToken cancellationToken)
+        {
+            var updateResult = new UpdateResult { ServerUpdated = false, ModsUpdated = false };
+
+            if (!System.Environment.Is64BitOperatingSystem)
+            {
+                var result = MessageBox.Show("ARK: Survival Evolved(tm) Server requires a 64-bit operating system to run.  Your operating system is 32-bit and therefore the Ark Server Manager will be unable to start the server, but you may still install it or load and save profiles and settings files for use on other machines.\r\n\r\nDo you wish to continue?", "64-bit OS Required", MessageBoxButton.YesNo, MessageBoxImage.Warning);
```
Comment (ChronosWS) [inline ARK Server Manager/Lib/Model/ServerRuntime.cs]: This needs to come from the localization file.  Were you able to get the run-time query of loc strings from XAML to work?

**Comment by Bletch1971** [comment]
Closing this and re-submitting as it can no longer be merged.

**Pull request closed without merging**